plugins/inputs/snmp/snmp.go

```diff
@@ -135,15 +135,20 @@ type Snmp struct {
 	Name   string
 	Fields []Field `toml:"field"`
 
-	connectionCache map[string]snmpConnection
+	sync.Mutex
+	connectionCache map[int]snmpConnection
 	initialized     bool
 }
 
 func (s *Snmp) init() error {
+	s.Lock()
+	defer s.Unlock()
 	if s.initialized {
 		return nil
 	}
 
+	s.connectionCache = make(map[int]snmpConnection)
```

Comment:
Since we're changing the cache key to be the agent's index in the list, we can turn this into a slice instead, using the index in `s.Agents` as the index for `s.connectionCache`. And then since we'd be using a slice, we can pre-allocate it here as we know the size of `s.Agents`. And then once we pre-allocate it, we can do away with the mutex since it'll never change during runtime. The only change will be to individual elements, which the updated description for `getConnection` says is an error to access in more than 1 goroutine.

None of that is required, just throwing it out there :-)

Reply:
Good point, I'll do this.